### ISSUE TITLE
Bwapi 4.4.0 support

### ISSUE THREAD

**Issue by Ravaelles** [basil-ladder/sc-docker#18]
I'm trying out this great project and bumped into an issue. Actually four of them 😂 

# Issue 1 - Bwapi 4.4.0 support

In the doc it's mentioned in one place that 4.4.0 is supported, in the other that it's not.

In `bwapi.py` there's no 4.4.0, so I've tried tweaking it to add 4.4.0 support like this:
```
versions_md5s = {
    '4.4.0': 'cf7a19fe79fad87f88177c6e327eaedc',
    '4.2.0': '2f6fb401c0dcf65925ee7ad34dc6414a',
    '4.1.2': '1364390d0aa085fba6ac11b7177797b0',
    '4.0.1b': '84f413409387ae80a4b4acc51fed3923',
    '3.7.5': '5e590ea55c2d3c66a36bf75537f8655a',
    '3.7.4': '6e940dc6acc76b6e459b39a9cdd466ae'
}
supported_versions = ['4.4.0', '4.2.0', '4.1.2', '3.7.5']
```

____

# Issue 2 - SSCAI dead means can't download map pack

With SSCAI being down, the `map.py` is broken as there are no maps to be downloaded, so for now I've uploaded it to my personal ftp:
`map.py:19`
```
download_extract_zip('https://ravaelles.usermd.net/sc/sscai_maps_pack.zip',
```

Perhaps it could be used for now? 

____

# Issue 3 - Can't say why my bot won't run

I was able to make the bwapi error disappear, but I'm still unable to run my 4.4.0 bot, it may be my fault (or may not), can't get any logs in GAMES dir. I have bot.json and other basic stuff set up.

When I run it with ` --log_level=DEBUG --log_verbose` it's the same, all I get is vague `some containers exited prematurely`:
```
2024-03-02 19:05:40 INFO scbw.docker_utils[26960] launched BotPlayer:Florian Richoux:P
2024-03-02 19:05:40 DEBUG scbw.docker_utils[26960] container name = 'GAME_2B634EDD_1_Florian_Richoux', container id = '['2e888dbecd']'
2024-03-02 19:05:40 DEBUG scbw.docker_utils[26960] checking if game has launched properly...
Traceback (most recent call last):
  File "D:\APPS\Python310\lib\site-packages\scbw\cli.py", line 231, in main
    game_result = run_game(args)
  File "D:\APPS\Python310\lib\site-packages\scbw\game.py", line 107, in run_game
    launch_game(players, launch_params, args.show_all, args.read_overwrite,
  File "D:\APPS\Python310\lib\site-packages\scbw\docker_utils.py", line 377, in launch_game
    raise DockerException(
scbw.error.DockerException: some containers exited prematurely, please check logs
```

It works fine for 2 players like "Florian Richoux" & "Florian Richoux". I get the replays etc.

___

# Question 4 - Possible to get bot binaries?

With SSCAIT being dead, I can't get a binary for any 4.4.0 bot. Would it be possible to make at least some of them publicly available? For now I'd love to get any .jar bot that uses 4.4.0 to verify.

**Comment by Bytekeeper** [comment]
### Regarding Issue 1
Are you sure you did check the code of this project and not of the original sc-docker?
Here's the file in this repo:
https://github.com/basil-ladder/sc-docker/blob/master/scbw/bwapi.py

This project is used extensively by the [BASIL ladder](https://www.basil-ladder.net/) which has 4.4.0 bots.

### Regarding Issue 2
Ah yeah, I might just upload the zip to this repo. The SSCAIT page is currently managed by a new maintainer, and might be back up in a few weeks.

### Regarding Issue 3
Your bot (I guess Atlantis with JBWAPI) should run, but maybe you used the original repo instead of my fork. Please note that the instructions for Windows are a bit different - and not tested/checked by me.

### Regarding Question 4
You can get some here: https://www.cs.mun.ca/~dchurchill/starcraftaicomp/2023/bots/
At least PurpleWave is JBWAPI bot.
Other than that: I did not run the SSCAIT page. I'm running the BASIL ladder, which also heavily relied on SSCAIT. While I do have all bot binaries, I cannot hand them out as some are not from SSCAIT, but from direct downloads. Since I did not ask the authors for permission to provide binaries, I won't.

A small question: Your framework Atlantis is actually well known in the bot community. Have you considered joining our [Discord](https://discord.gg/xHC2Jf9SQG)? I also always wondered why you did not submit an actual bot to SSCAIT?

**Comment by Ravaelles** [comment]
I can't say how much I appreciate your fast & thorough answer. Also thank you for the incentive to join the Discord and mentioning Atlantis. The reason why I didn't join SSCAIT calls for its own book. To make a very long story short: mixture of psychological reasons (_I'm not good enough yet_) and the fact that I was battling all sort of pesky bugs, with very limited success. Most them caused by programming being dead tired and doing false instead of true etc.
Furthermore I got mental breakdown thanks to compulsively writing it. Twice. And get seriously addicted 3 times, quit it, get back and so on. However I still have in plan fixing all the bugs and uploading Atlantis :- )
I have it mind for the past 6 years :- )

My ego was hurt beyond words by doing all of this, but it made me a better person. Like, really. By showing me how weak of a programmer I was. A cruel & yet valuable lesson.

Getting back to running sc-docker:
- I realized I was having old images from somewhere around ~2yrs ago,
- so I re-installed fresh python3.6 and scbw,
- It still didn't work, my `bwapi.py` with full fresh install didn't support 4.4.0,
- so I downloaded manually this repo & copied files into the `scbw` dir, only then it worked.
- (Also had to manually fix the map pack .zip that I mentioned earlier)
- Thanks a lot for Dave's bots too!

**Comment by Bytekeeper** [comment]
I understand where you're coming from. Imposter syndrome is pretty common in bot devs, due to the literal monsters at the top.

It's not "Dave's" bots - those are the bots for the last tournament he organized. But I guess you know that - and it was a misunderstanding on my part.

As for bwapi.py etc. I can only assume you checked out the original. Also, you really really need to build the docker files yourself. Using the ones on Dockerhub will most likely not work with my fork - or at least not fully.

Another way to fire up games vs other bots is using [BWAIShotgun](https://github.com/Bytekeeper/BWAIShotgun). On Windows it might be the easier alternative, but the amount of games that can run parallel is more limited.

**Comment by Ravaelles** [comment]
Yup, I meant "_bots on Dave's server_".

About building the images: I forgot to mention that this time I ran `./build_images.ps1`. Perhaps that was the missing step that I ignored yesterday (assuming that I already had the images).

Again, thanks a lot for your abso-freakin-lutely great work on these projects.

I will definitely try out BWAIShotgun today, funny that I haven't heard about it before. I expect I'll get back to you for some more help 😂 ...or to report (unlikely) success. I'm really interested in running the bot as quickly as possible, in as few clicks as possible, so it sounds like a great tool! ❤️
